src/app/icd/client/CheckInDelegate.h

```diff
@@ -39,6 +42,62 @@ class DLL_EXPORT CheckInDelegate
                                node that sent the check-in message.
      */
     virtual void OnCheckInComplete(const ICDClientInfo & clientInfo) = 0;
+
+    /**
+     * @brief Callback used to let the application know that a key refresh is
+     * needed to avoid counter rollover problems.
+     *
+     * The implementer of this function should dynamically create a new ICDRefreshKeyInfo object. They should also generate a new
+       key and store it in the ICDRefreshKeyInfo object.
+       The implementer of this function should use CreateICDRefreshKeyInfo() to create the object
+     *
+     * @param[in] clientInfo - ICDClientInfo object representing the state associated with the
+                               node that sent the check-in message. The callee can use the clientInfo to determine the type of key
+                               to generate.
+     * @param[in] clientStorage - ICDClientStorage object stores the updated ICDClientInfo after re-registration into
+                                  persistent storage.
+     * @return ICDRefreshKeyInfo - pointer to the ICDRefreshKeyInfo object comprising the newly generated key, the input
+                                  ICDClientInfo, pointer to the CheckInDelegate and CommandSender object for re-registering with the
+                                  new key.
+    */
+    virtual ICDRefreshKeyInfo * OnKeyRefreshNeeded(const ICDClientInfo & clientInfo, const ICDClientStorage * clientStorage) = 0;
+
+    /**
+     * @brief Callback used to let the application know that the re-registration with the new key was successful and provides the
+     * updated ICDClientInfo
+     *
+     * @param[in] clientInfo - ICDClientInfo object representing the state associated with the
+                               node that sent the check-in message. This will have the new key used for registration and the updated
+                               icd counter. The application should clean up the ICDRefreshKeyInfo object using
+                               ReleaseICDRefreshKeyInfo() on receipt.
```

Comment:
Again, ReleaseICDRefreshKeyInfo() seems like implementation detail that is not relevant to this API.  The important part is that after this call the caller will no longer use the `ICDRefreshKeyInfo *` that it hands in here.

Reply:
Updated the description.